safehttp/doc.go

```diff
@@ -12,5 +12,113 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package safehttp TODO(@kele, @empijei)
+// Package safehttp provides a framework for building secure-by-default HTTP
+// servers. See https://github.com/google/go-safeweb#readme to learn about the
+// goals and features.
+//
+// Safe Responses
+//
+// HTTP responses need to be crafted carefully in order to prevent common web
+// vulnerabilities like Cross-site Scripting (XSS). To help with this, we use
+// Safe Responses.
+//
+// Safe Responses are HTTP responses which has been determined to be safe when
+// received by a modern, popular web browser. For example, we consider HTML
+// responses generated using a template system enforcing contextual autoescaping
+// to be safe e.g. modern Angular, github.com/google/safehtml/template).
+//
+// The Go Safe Web ResponseWriter implementation accepts exclusively Safe
+// Responses, instead of byte slices.
+//
+// Since different projects will consider different ways of crafting a response
+// safe, we offer a way of configuring this in the framework. Whether a response
+// is considered safe or not is determined by the Dispatcher.
+//
+// Dispatcher
+//
+// An implementation of a Dispatcher should be provided by security experts in
+// your project. The Dispatcher is called for every write method of the
+// ResponseWriter and is used to determine whether the passed response should be
+// considered safe. The Dispatcher is responsible for writing the response to
+// the underlying http.ResponseWriter in a safe way.
+//
+// Go Safe Web provides a DefaultDispatcher implementation which supports
+// github.com/google/safehtml responses.
+//
+// Interceptors
+//
+// Not all security features can be implemented using the Dispatcher alone. For
+// instance, some requests should be rejected before they reach the handler in
+// order to prevent from Cross-site Request Forgery (CSRF, XSRF). To support
+// this, the framework uses Interceptors.
+//
+// An Interceptor implements methods that run before the request is passed to
+// the handler, and after the handler has committed a response. These are,
+// respectively, Before and After.
+//
+// Life of a Request
+//
+// To tie it all together, we will explain how a single request goes through the
+// framework.
+//
+// When the request reaches the server, the following happens:
+//
+// 1. The ServeMux routes the request (i.e. picks the appropriate Handler that
+// will be eventually called). The HTTP method of the request is checked whether
+// it matches any registered Handler.
+//
+// 2. [Before phase] The request is passed to all installed Interceptors, to
```

Comment:
The requestis passed to the Before methods of all installed Interceptors in the order of their installation on the ServeMux.